### ISSUE TITLE
fluid.layers.l2_normalize output nan when using gpu mode

### ISSUE THREAD

**Issue by kebinC** [PaddlePaddle/Paddle#13235]
fluid.layers.l2_normalize work well when using cpu mode, but it output nan changing to gpu mode

when i replace "fluid.layers.l2_normalize" by 
`    weight_norm = fluid.layers.sqrt(fluid.layers.reduce_sum(fluid.layers.square(weight), dim=1))
    weight = fluid.layers.elementwise_div(weight, weight_norm, axis=0)`,
it alse work well.

```
import os
import sys
import math
import time
import argparse
import functools
import numpy as np
import paddle
import paddle.fluid as fluid


def arc_margin_product(input, label, out_dim, m=0.2, s=20.0):
    input = fluid.layers.l2_normalize(input, axis=1)
    weight = fluid.layers.create_parameter(
                shape=[out_dim, input.shape[1]],
                dtype='float32',
                name='weight_norm',
                attr=fluid.param_attr.ParamAttr(
                              initializer=fluid.initializer.Xavier()))
    weight = fluid.layers.l2_normalize(weight, axis=1)
    #weight_norm = fluid.layers.sqrt(fluid.layers.reduce_sum(fluid.layers.square(weight), dim=1))
    #weight = fluid.layers.elementwise_div(weight, weight_norm, axis=0)
    weight = fluid.layers.transpose(weight, perm = [1, 0])
    cosine = fluid.layers.mul(input, weight)
    return cosine


def convolutional_neural_network(img, label):
    conv_pool_1 = fluid.nets.simple_img_conv_pool(
        input=img,
        filter_size=5,
        num_filters=20,
        pool_size=2,
        pool_stride=2,
        act="relu")
    conv_pool_1 = fluid.layers.batch_norm(conv_pool_1)
    conv_pool_2 = fluid.nets.simple_img_conv_pool(
        input=conv_pool_1,
        filter_size=5,
        num_filters=50,
        pool_size=2,
        pool_stride=2,
        act="relu")
    fc_1 = fluid.layers.fc(input=conv_pool_2, size=2048)
    prediction = arc_margin_product(fc_1, label, 2048)
    return prediction

def train():
    image = fluid.layers.data(name='image', shape=[1, 28, 28], dtype='float32')
    label = fluid.layers.data(name='label', shape=[1], dtype='int64')

    # model definition
    out = convolutional_neural_network(image, label)

    test_program = fluid.default_main_program().clone(for_test=True)


    #place = fluid.CUDAPlace(0)
    place = fluid.CPUPlace()
    exe = fluid.Executor(place)
    exe.run(fluid.default_startup_program())

    fetch_list_train = [out.name]

    data = np.random.uniform(0, 1, (10, 1, 28, 28)).astype(np.float32)
    label = np.arange(10).astype(np.int64).reshape(10, 1)
    loss = exe.run(test_program, fetch_list=fetch_list_train, feed={'image': data, 'label':label})
    print loss

def main():
    train()

if __name__ == '__main__':
    main()

```

**Comment by luameows** [comment]
我也遇到这个问题，但是没有尝试cpu下的情况。
另外，我才用`elementwise_div`也变成nan。
`fluid.layers.create_parameter`生成的参数同样是nan。

**Comment by luotao1** [comment]
related https://github.com/PaddlePaddle/Paddle/issues/13200#issuecomment-418363398

**Comment by qingqing01** [comment]
@luameows 你可以尝试用 类似的方法做l2 norm:

```
weight_norm = fluid.layers.sqrt(fluid.layers.reduce_sum(fluid.layers.square(weight), dim=1)) 
weight = fluid.layers.elementwise_div(weight, weight_norm, axis=0),
```

CPU和GPU内部是同一套实现，很奇怪CPU正确，GPU出错，这个问题我们看下。多谢了~

**Comment by luameows** [comment]
@qingqing01 我就是采用这样的方法做的。
在我的例程里，input_x是网络输出特征。我开始尝试利用l2 norm对其进行归一化，出现nan，后来采用下述方法，最后输出input_x_norm也是nan，然后我跟进发现input_x_sqrt及其之上的所有变量的值均存在，在调用`elementwise_div`之后就变成`nan`。
```
input_x_square=fluid.layers.square(x=input_x)
input_x_sum=fluid.layers.reduce_sum(input=input_x_square,dim=1,keep_dim=True)
input_x_sqrt=fluid.layers.sqrt(x=input_x_sum)
input_x_norm=fluid.layers.elementwise_div(x=out1,y=input_x_sqrt,axis=0)
```

**Comment by qingqing01** [comment]
在div之前加个epsion试下: 
```
input_x_square=fluid.layers.square(x=input_x)
input_x_sum=fluid.layers.reduce_sum(input=input_x_square,dim=1,keep_dim=True)
input_x_sqrt=fluid.layers.sqrt(x=input_x_sum)
input_x_sqrt =  input_x_sqrt + 1e-6  # 1e-4
input_x_norm=fluid.layers.elementwise_div(x=out1,y=input_x_sqrt,axis=0)
```
另外，l2_norm也有个epsilon可调节试下~

**Comment by luameows** [comment]
@qingqing01 这又是一个奇怪的地方了，你说的这段代码我也添加过。但是添加之后，`input_x_sqrt`就变成了nan

**Comment by luameows** [comment]
@qingqing01 有什么命令可以查看当前paddle版本么？我怀疑是paddle版本的问题吧，之前看到issuse中有修复过l2 norm的bug

**Comment by luameows** [comment]
我在teamcity下载了最新版whl包，安装后，l2_norm还是有问题

**Comment by qingqing01** [comment]
@luameows @kebinC  测试发现可能是Eigen表达式在GPU时出错，我提交了一个PR，重新用CUDA实现了下~

**Comment by luameows** [comment]
@qingqing01 好的，我之后更新测试下。能不能帮忙看下 #13276  我最新测试的情况？

**Comment by luameows** [comment]
@qingqing01 新版本测试还是有问题。
我测试了feature的值都是正值，通过`l2_norm`之后得到的结果竟然有负值出现。

**Comment by qingqing01** [comment]
@luameows  可否给我一个最简单的复现代码，我测下~
